pkg/kubelet/kubelet_pods.go

```diff
@@ -2420,3 +2443,55 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
 		go pcm.Destroy(val)
 	}
 }
+
+func (kl *Kubelet) runtimeClassSupportsRecursiveReadOnlyMounts(pod *v1.Pod) bool {
+	var runtimeClassName string
+	if pod.Spec.RuntimeClassName != nil {
+		runtimeClassName = *pod.Spec.RuntimeClassName
+	}
+	runtimeHandlers := kl.runtimeState.runtimeHandlers()
+	return runtimeClassSupportsRecursiveReadOnlyMounts(runtimeClassName, runtimeHandlers)
+}
+
+// runtimeClassSupportsRecursiveReadOnlyMounts checks whether the runtime class supports recursive read-only mounts.
+// The kubelet feature gate is not checked here.
+func runtimeClassSupportsRecursiveReadOnlyMounts(runtimeClassName string, runtimeHandlers []kubecontainer.RuntimeHandler) bool {
+	for _, h := range runtimeHandlers {
+		if h.Name == runtimeClassName {
+			return h.SupportsRecursiveReadOnlyMounts
+		}
+	}
+	klog.ErrorS(nil, "unknown runtime class", "runtimeClassName", runtimeClassName)
+	return false
+}
+
+// resolveRecursiveReadOnly resolves the recursive read-only mount mode.
+func resolveRecursiveReadOnly(m v1.VolumeMount, runtimeSupportsRRO bool) (bool, error) {
+	if m.RecursiveReadOnly == nil {
+		return false, nil
+	}
+	switch rroMode := *m.RecursiveReadOnly; rroMode {
+	case v1.RecursiveReadOnlyDisabled:
+		return false, nil
+	case v1.RecursiveReadOnlyIfPossible:
+		propNone := true
+		if m.MountPropagation != nil {
+			propNone = *m.MountPropagation == v1.MountPropagationNone
+		}
+		return runtimeSupportsRRO && m.ReadOnly && propNone, nil
```

Comment:
I still think that we should return errors in case of readonly true or mountProp not set to None because those fields are in control of the user writing the pod spec. If their runtime actually supports RRO they would miss it when we have an opportunity to return an error so they can correct it and get RRO.

Reply:
fixed